### PR TITLE
Make event registration non-blocking for clients

### PR DESCRIPTION
**Pull request by ihsandemir**
Changed the event registration and de-registration interfaces to work with async futures. This allows the callers to choose between async and sync versions depending on the need. All the client message tasks will use the async version while the server side proxies sync on the future before returning to the caller.

Partial fix for https://github.com/hazelcast/hazelcast/issues/15085

### CODE REVIEW

**Comment by ihsandemir** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/cache/impl/ICacheService.java [hunk 1]
```diff
@@ -119,7 +123,8 @@
      */
     CacheOperationProvider getCacheOperationProvider(String cacheNameWithPrefix, InMemoryFormat storageType);
 
-    UUID addInvalidationListener(String cacheNameWithPrefix, CacheEventListener listener, boolean localOnly);
+    CompletableFuture<EventRegistration> addInvalidationListener(String cacheNameWithPrefix, CacheEventListener listener,
```
Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/ICacheService.java]: Why not return CompetableFuture<UUId> or even Future<UUID> ? They were returning UUID before, then it should be sufficient tor return ` Future<UUID>` for the async versions.

File: hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java [hunk 2]
```diff
@@ -574,52 +575,60 @@ public void dispatchEvent(Object event, CacheEventListener listener) {
...
         }
 
-        UUID id = reg.getId();
-        if (listener instanceof Closeable) {
-            closeableListeners.put(id, (Closeable) listener);
-        } else if (listener instanceof CacheEntryListenerProvider) {
-            CacheEntryListener cacheEntryListener = ((CacheEntryListenerProvider) listener).getCacheEntryListener();
-            if (cacheEntryListener instanceof Closeable) {
-                closeableListeners.put(id, (Closeable) cacheEntryListener);
+        registrationFuture.whenCompleteAsync((BiConsumerEx<EventRegistration, Throwable>) (eventRegistration, throwable) -> {
```
Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java]: Cause of the failure https://github.com/hazelcast/hazelcast/issues/15935
We should return the returned complationStage from `whenCompleteAsync` method.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/RegistrationUtil.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+package com.hazelcast.spi.impl.eventservice.impl;
+
+import javax.annotation.Nonnull;
+import java.util.UUID;
+import java.util.concurrent.Future;
+
+import static com.hazelcast.internal.util.ExceptionUtil.rethrow;
+
+public interface RegistrationUtil {
```
Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/RegistrationUtil.java]: This two methods are doing the same thing. I would merge them into one function with generics and also put it under FutureUtil instead of here. Also InterruptedException should be better handled. 
```
 @Nonnull
    static <T> T join(Future<T> future) {
        try {
            return future.get();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw rethrow(e);
        } catch (ExecutionException e) {
            throw rethrow(e);
        }
    }
```

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractListenerMessageTask.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import com.hazelcast.instance.impl.Node;
+import com.hazelcast.internal.nio.Connection;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+
+/**
+ * Base message task for listener registration tasks
+ */
+public abstract class AbstractListenerMessageTask<P, T>
```
Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractListenerMessageTask.java]: This class has no specific logic to Listeners. It becomes more clear when you make `eventRegistrationFuture`  a local variable. I think removing it will make both AbstractAddListenerMessageTask and AbstractRemoveListenerMessageTask.

Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractListenerMessageTask.java]: Since it has some simplification because of generics. We may rename it to something more generic. 
`AbstractAsyncMessageTask` ?
AbstractAddress and AbstractInvocationMessageTask can extend this class. I see similar pattern there also. 
Also some usages has optimizations while others are missing. 
See 

```
 @Override
    public void execute(Runnable command) {
        if (Thread.currentThread().getClass() == PartitionOperationThread.class) {
            // instead of offloading it to another thread, we run on the partition thread. This will speed up throughput.
            command.run();
        } else {
            ExecutionService executionService = nodeEngine.getExecutionService();
            Executor executor = executionService.getExecutor(ExecutionService.ASYNC_EXECUTOR);
            executor.execute(command);
        }
    }

    @Override
    public void accept(Object response, Throwable throwable) {
        if (throwable == null) {
            beforeResponse();
            sendResponse(response);
            afterResponse();
        } else {
            beforeResponse();
            handleProcessingFailure(throwable);
            afterResponse();
        }
    }
```

File: hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/EventService.java [hunk 5]
```diff
@@ -58,12 +60,11 @@
...
      * @param topic       topic name
      * @param listener    listener instance
-     * @return event registration
+     * @return event registration future
      * @throws IllegalArgumentException if the listener is {@code null}
      */
-    EventRegistration registerLocalListener(@Nonnull String serviceName,
-                                            @Nonnull String topic,
-                                            @Nonnull Object listener);
+    CompletableFuture<EventRegistration> registerLocalListener(@Nonnull String serviceName, @Nonnull String topic,
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/EventService.java]: i think, it can be better not to introduce unneeded functionality, async local calls are not required as far as i see.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/EventService.java [hunk 6]
```diff
@@ -72,26 +73,23 @@ EventRegistration registerLocalListener(@Nonnull String serviceName,
...
      * @param topic       topic name
      * @param listener    listener instance
-     * @return event registration
+     * @return event registration future
      * @throws IllegalArgumentException if the listener is {@code null}
      */
-    EventRegistration registerListener(@Nonnull String serviceName,
-                                       @Nonnull String topic,
-                                       @Nonnull Object listener);
+    CompletableFuture<EventRegistration> registerListener(@Nonnull String serviceName, @Nonnull String topic,
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/EventService.java]: what about introducing an async variant besides sync one? This can lead less change and make the places which we need async calls clear.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAddListenerMessageTask.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.instance.impl.Node;
+import com.hazelcast.internal.nio.Connection;
+
+import java.util.UUID;
+
+/**
+ * Base message task for listener registration tasks
+ */
+public abstract class AbstractAddListenerMessageTask<P>
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAddListenerMessageTask.java]: can be renamed as `AbstractDestroyableListenerMessageTask`?

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAsyncMessageTask.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+    /**
+     * Called on node side, after process is run and right before sending the response to the client.
+     *
+     * This method gives the task a chance to change the response object to be sent to the client
+     *
+     * @param response The response for the task. It can be modified to a different object by overriding this method.
+     * @return The object to be sent to the client. The user may choose to return a different
+     */
+    protected Object beforeResponse(T response) {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAsyncMessageTask.java]: `modifyResponse`?

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAsyncMessageTask.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        return response;
+    }
+
+    /**
+     * Called on node side, after sending the response to the client.
+     *
+     * @param response  The response sent to the client
+     * @param throwable The throwable to be sent to the client if an exception occured
+     */
+    protected void afterResponse(Object response, Throwable throwable) {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAsyncMessageTask.java]: `afterSendingResponse`?

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAsyncMessageTask.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Override
+    protected void processMessage() {
+        beforeProcess();
+        CompletableFuture<T> internalFuture = processInternal();
+        internalFuture.thenApply(this::beforeResponse).whenComplete(this::handleResponse).whenComplete(this::afterResponse);
+    }
+
+    private void handleResponse(Object response, Throwable throwable) {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAsyncMessageTask.java]: `sendResponse`?

File: hazelcast/src/main/java/com/hazelcast/internal/util/InvocationUtil.java [hunk 11]
```diff
@@ -67,7 +67,8 @@ private InvocationUtil() {
 
         ClusterService clusterService = nodeEngine.getClusterService();
         if (!clusterService.isJoined()) {
-            return newCompletedFuture(null);
+            // If this node did not join the cluster yet, invocation to the cluster is not possible.
+            return newCompletedFuture(true);
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/internal/util/InvocationUtil.java]: is this change required for this PR?

File: hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/operations/RegistrationOperation.java [hunk 12]
```diff
@@ -40,12 +39,12 @@ public RegistrationOperation(Registration registration, int memberListVersion) {
...
     protected void runInternal() {
         EventServiceImpl eventService = (EventServiceImpl) getNodeEngine().getEventService();
-        response = eventService.handleRegistration(registration);
+        eventService.handleRegistration(registration);
     }
 
     @Override
     public Object getResponse() {
-        return response;
+        return true;
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/operations/RegistrationOperation.java]: seems `false` is also a valid response.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/EventServiceImpl.java [hunk 13]
```diff
@@ -246,69 +247,105 @@ private long eventsProcessed() {
...
+     * Registers the listener for events matching the service name, topic and filter on local member.
+     *
+     * @param serviceName the service name for which we are registering
+     * @param topic       the event topic for which we are registering
+     * @param filter      the filter for the listened events
+     * @param listener    the event listener
+     * @return the event registration
+     * @throws IllegalArgumentException if the listener or filter is null
+     */
+    private EventRegistration registerListenerInternalLocally(@Nonnull String serviceName, @Nonnull String topic,
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/EventServiceImpl.java]: maybe we can call this just `registerListener`?

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAddListenerMessageTask.java (hunk 7 above)
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAddListenerMessageTask.java]: It adds listener so existing name sounds better.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAsyncMessageTask.java (hunk 8 above)
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAsyncMessageTask.java]: what about a name like: `processResponseBeforeSending` ?

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAsyncMessageTask.java (hunk 10 above)
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAsyncMessageTask.java]: Changed to `sendResponseOrHandleFailure`

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAsyncMessageTask.java (hunk 8 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAsyncMessageTask.java]: sounds more clear

File: hazelcast/src/main/java/com/hazelcast/internal/util/InvocationUtil.java (hunk 11 above)
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/internal/util/InvocationUtil.java]: reverted.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/EventServiceImpl.java [hunk 14]
```diff
@@ -253,7 +253,7 @@ public EventRegistration registerLocalListener(@Nonnull String serviceName,
     @Override
     public EventRegistration registerLocalListener(@Nonnull String serviceName, @Nonnull String topic,
                                                    @Nonnull EventFilter filter, @Nonnull Object listener) {
-        return registerLocalListenerInternal(serviceName, topic, filter, listener, true);
+        return registerListenerInternalLocally(serviceName, topic, filter, listener, true);
```
[the remark this answers is not in the document]
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/EventServiceImpl.java]: Changed to `registerListener0`

File: hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/operations/RegistrationOperation.java (hunk 12 above)
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/operations/RegistrationOperation.java]: good catch, reverted the change.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/EventServiceImpl.java (hunk 13 above)
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/EventServiceImpl.java]: See https://github.com/hazelcast/hazelcast/pull/15860#discussion_r347896293

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAsyncMessageTask.java (hunk 8 above)
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractAsyncMessageTask.java]: changed to `processResponseBeforeSending `.
